src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+
+/* Poll the server monitor stream for reading. Allows cancellation.
+ *
+ * Called only from server monitor thread.
+ * Locks server monitor mutex.
+ * Returns true if stream is readable. False on error or cancellation.
+ * On cancellation, no error is set, but cancelled is set to true.
+ */
+static bool
+_server_monitor_poll_with_interrupt (mongoc_server_monitor_t *server_monitor,
+                                     int64_t expire_at_ms,
+                                     bool *cancelled,
+                                     bson_error_t *error)
+{
+   /* How many milliseconds we should poll for on each tick.
+    * On every tick, check whether the awaitable ismaster was cancelled. */
+   const int32_t k_monitor_tick_ms = 500;
+   int64_t timeleft_ms;
+
+   MONITOR_LOG (server_monitor,
+                "_server_monitor_poll_with_interrupt expire_at_ms: %" PRIu64,
+                expire_at_ms);
+   while ((timeleft_ms = expire_at_ms - _now_ms ()) > 0) {
+      ssize_t ret;
+      mongoc_stream_poll_t poller[1];
+
+      poller[0].stream = server_monitor->stream;
+      poller[0].events =
+         POLLIN; /* POLLERR and POLLHUP are added in mongoc_socket_poll. */
+      poller[0].revents = 0;
+
+      MONITOR_LOG (
+         server_monitor,
+         "polling for awaitable ismaster reply with timeleft_ms: %" PRId64,
+         timeleft_ms);
+      ret = mongoc_stream_poll (
+         poller, 1, (int32_t) BSON_MIN (timeleft_ms, k_monitor_tick_ms));
+      if (ret == -1) {
+         MONITOR_LOG (server_monitor, "mongoc_stream_poll error");
+         bson_set_error (error,
+                         MONGOC_ERROR_STREAM,
+                         MONGOC_ERROR_STREAM_SOCKET,
+                         "poll error");
+         return false;
+      }
+
+      if (poller[0].revents & (POLLERR | POLLHUP)) {
+         bson_set_error (error,
+                         MONGOC_ERROR_STREAM,
+                         MONGOC_ERROR_STREAM_SOCKET,
+                         "connection closed while polling");
+         return false;
+      }
+
+      /* Check for cancellation. */
+      bson_mutex_lock (&server_monitor->shared.mutex);
+      *cancelled = server_monitor->shared.cancel_requested;
+      server_monitor->shared.cancel_requested = false;
+      bson_mutex_unlock (&server_monitor->shared.mutex);
+
+      if (*cancelled) {
+         MONITOR_LOG (server_monitor, "polling cancelled");
+         return false;
+      }
+
+      if (poller[0].revents & POLLIN) {
+         MONITOR_LOG (server_monitor, "mongoc_stream_poll ready to read");
+         return true;
+      }
+   }
+   bson_set_error (error,
+                   MONGOC_ERROR_STREAM,
+                   MONGOC_ERROR_STREAM_SOCKET,
+                   "connection timeout while polling");
+   return false;
+}
+
+/* Receive an awaitable ismaster reply.
+ *
+ * May be used to receive additional replies when moreToCome is set.
+ * Called only from server monitor thread.
+ * Locks server monitor mutex.
```

Comment:
I don't see `bson_mutex_lock` called in this function. Is it called indirectly through some other function?

Reply:
Clarification seems fine. Thanks.